inst/@sym/dsolve.m

```diff
@@ -180,13 +167,43 @@
   end
 
   cmd = { 'ode=_ins[0]; ics=_ins[1:]'
-          '# convert our input to a dict'
-          'ics2 = {}'
-          'for s in ics:'
-          '    ics2[s.lhs] = s.rhs'
-          'sol = sp.dsolve(ode, ics=ics2)'
-          'return sol,' };
-  soln = pycall_sympy__ (cmd, ode, varargin{:});
+          'ics = {ic.lhs: ic.rhs for ic in ics}'
+          'sol = sp.dsolve(ode, ics=ics)'
+          % Helper function
+          'def convert(sympy_obj):'
+          '    if isinstance(sympy_obj, Eq) and sympy_obj.lhs.is_Function:'
+                   % type() of a sympy function gives an undefined sympy function, I guess
+                   % str to turn it into just the string like matlab
+          '        return str(type(sympy_obj.lhs)), sympy_obj.rhs'
+          '    return None'
+          % If the solution set is iterable (system or multiple solutions),
+          % we will convert it to a structure of {eqname: expr, eqname2: expr2 ...}
+          % 'cause that's what matlab does!
+          'try:'
+          '    iterator = iter(sol)'
+          'except TypeError:'
+          % Not iterable
+          '    pass'
+          'else:'
+          % Iterable
+          '    return_data = dict()'
+          '    for solution_part in sol:'
+          '        results = convert(solution_part)'
+          '        if results is None:'
+          '            return 0, -1'
+          '        return_data[results[0]] = results[1]'
```

Comment:
I'm concerned here if the thing we iterate over could have multiple solutions then maybe two things in the iterator end up with key `"y"` (say).  In this case the 2nd time would overwrite the first one.